src/erc721/ERC721.sol

```diff
@@ -0,0 +1,237 @@
+    
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool supported) {
+        supported = interfaceId == 0x80ac58cd || interfaceId == 0x5b5e139f;
+    }
+    
+    function approve(address spender, uint256 tokenId) external {
+        address owner = ownerOf[tokenId];
+        
+        require(msg.sender == owner || isApprovedForAll[owner][msg.sender], "NOT_APPROVED");
+        
+        getApproved[tokenId] = spender;
+        
+        emit Approval(owner, spender, tokenId); 
+    }
+    
+    function setApprovalForAll(address operator, bool approved) external {
+        isApprovedForAll[msg.sender][operator] = approved;
+        
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function transferFrom(address, address to, uint256 tokenId) public {
+        address owner = ownerOf[tokenId];
+        
+        require(
+            msg.sender == owner 
+            || msg.sender == getApproved[tokenId]
+            || isApprovedForAll[owner][msg.sender], 
+            "NOT_APPROVED"
+        );
+        
+        // This is safe because ownership is checked
+        // against decrement, and sum of all user
+        // balances can't exceed type(uint256).max!
+        unchecked { 
+            balanceOf[owner]--; 
+        
+            balanceOf[to]++;
+        }
+        
+        delete getApproved[tokenId];
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(owner, to, tokenId); 
+    }
+    
+    function safetransferFrom(address, address to, uint256 tokenId) external {
+        safetransferFrom(address(0), to, tokenId, "");
+    }
+    
+    function safetransferFrom(address, address to, uint256 tokenId, bytes memory data) public {
+        transferFrom(address(0), to, tokenId); 
+        
+        if (to.code.length != 0) {
+            // selector = `onERC721Received(address,address,uint,bytes)`
+            (, bytes memory returned) = to.staticcall(abi.encodeWithSelector(0x150b7a02,
+                msg.sender, address(0), tokenId, data));
+                
+            bytes4 selector = abi.decode(returned, (bytes4));
+            
+            require(selector == 0x150b7a02, "NOT_ERC721_RECEIVER");
+        }
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                          PERMIT/EIP-2612-LIKE LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function permit(
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        require(deadline >= block.timestamp, "PERMIT_DEADLINE_EXPIRED");
+        
+        address owner = ownerOf[tokenId];
+        
+        bytes32 digest = keccak256(
+            abi.encodePacked(
+                "\x19\x01",
+                DOMAIN_SEPARATOR,
+                keccak256(abi.encode(PERMIT_TYPEHASH, spender, tokenId, nonces[owner]++, deadline))
```

Comment:
For NFTs, the nonce should probably be more `mapping(uint256 => uint256)` because it's for a tokenId.

Your implementation will invalidate any permit on any other id of this owner, once a permit is used.

Whenever we should here invalidate any other permit on this tokenId, not from this owner.

Reply:
good catch @dievardump :~)!